### ISSUE TITLE
Parameterized questions?

### ISSUE THREAD

**Issue by pablo-angulo** [nennigb/amc2moodle#13]
Do you plan to allow for parameterized questions with fpeval?

Bennoit Nennig proposed some possible approaches in this thread:

https://project.auto-multiple-choice.net/boards/2/topics/4224

**Comment by nennigb** [comment]
Hi,

We will work on that. I have started to explore few ideas...
I hope we could have a first version for testing, in few weeks.

Benoit

**Comment by nennigb** [comment]
I start to work on conversion of moodle calculated question to AMC. In the question text, Moodle encode the variable name as `{x}` and the computation as `{=2*sin({x})}`. The values generated by moodle are also stored in the XML file in `<dataset_definitions>` node.

To do it in a clean way, I use [Pyparsing](https://pyparsing-docs.readthedocs.io/en/latest/HowToUsePyparsing.html) **to extract the expressions and variables** from the `<questiontext>` and `<answernode>` nodes. It will be more verbose than using regex but we can hope better results at least on clarity.

 
The second steps, will be **to render them into latex**. Several rendering strategies will be available : `fp`, `pgfmathparse` or a loop over the values stored in XML file to generate several questions. My idea is to add support for all of them at the end.
Moodle mathematical built-in functions are more numerous than those of `fp`, thus the loop approach could be useful to properly render unsupported function using python...
For supported function, the XML dataset will be use to pass the good bounds to a random number generator like `\FPrandom`.

I also join a basic example of XML rename in txt for upload.

The conversion from AMC to moodle will be relatively similar with an additional step in LaTeXML. Again, we could chose between a loop that will map the AMC question to standard question moodle (without computation) or calculated question by mapping `fp` to moodle built-in functions.

[quiz-calculated.txt](https://github.com/nennigb/amc2moodle/files/4942008/quiz-calculated.txt)

**Comment by nennigb** [comment]
A first **beta** version is now available on the branch `feature-moodle2amc-calculated`.
For now, I test with `numerical.xml` present in the `moodle2amc/test` folder.
I still need to finalize automated tests in `calculatedParser.py` and `test.py`.
The 2 mains limitations are : latex variable names are limited to alphas, some functions are not supported by `fp`.

  1. Have you more weird examples to submit ? 
  2. Have you some remarks or did you find some other weakness ?

**Comment by nennigb** [comment]
A support for **AMC parametrized questions**, with `fp`, is now also available for testing in the branch `feature-amc2moodle-calculated`.  When `fp` is used to create a random parameter, the question will be converted and maped into moodle `CalculatedMulti` [question](https://docs.moodle.org/39/en/Calculated_question_type#Available%20functions) .
It is noteworthy that moodle calculated question can also be converted in AMC with `moodle2amc`.

Currently, only the main commands of `fp` are supported : `\FPset`, `\FPrandom`, `\FPseed` (ignored), `\FPpi`, `\FPeval` and `\FPprint` (other commands like `\FPadd` can be easily defined in `fp.sty.ltxml` using `\FPeval`).
Some `\FPeval` functions, like `trunc` and rounding functions are still missing. Most of them could be added. 

## Strategy
1. In the LaTeXML binding `fp.sty.ltxml`
  -  `\FPeval` expressions are not evaluated, but are used to define macro. For instance
```
\FPeval{\x}{2*3+sin(1/2)}
\FPeval{\y}{2*\x} - > '2*(2*3+sin(1/2))'
```
  -  `\FPprint` make substitution of all macro, especially those created with `\FPeval`, and create raw entries in the XML file. **Using this command is mandatory**.
  -  Only random variable created with '\FPrandom` or with `random` in `\FPeval` cannot be substituted. Each call of these commands will create a `rand` variable with a unique label. 

2. In `convert.py`
  - if  `\FPprint` is used in a question or in questionmult, a `CalculatedMulti` question will be created
  - the `rand1`,`rand2` ... variables are assign to a moodle wildcards.
  - the `fp` raw expressions from XML are parsed (with `pyparsing`) and recast to follow moodle [syntax](https://docs.moodle.org/39/en/Calculated_question_type#Available%20functions).
  - the dataset are created using python random generator (ie results will be different from those obtained by `fp`). 

## Example
```
\FPeval{\a}{1+random*(10-1)} % uniform in [1, 10]
\FPeval{\b}{2.5+3}
```
Moodle expects that the answer will contains only mathematical expression thus `amc2moodle` will also expect `choices` environment defined like
```
\begin{choices}
   \correctchoice{\FPprint{\FPeval{\out}{\a * \b}\out}}
   \wrongchoice{\FPprint{\FPeval{\out}{\a + \b}\out}}
\end{choices}
``` 
It is noteworthy that there is NO `$x=$` or additional things in the choices...

More example are available in the [test suite](https://github.com/nennigb/amc2moodle/blob/feature-amc2moodle-calculated/amc2moodle/amc2moodle/test/common-bank.tex#L195)

## Limitations 
  - moodle wildcards, cannot be replaced in equation rendered by mathjax.
  - Only private datasets are currently supported.

## TO DO
  - add magic comment to pass options (dataset size, ...)
  - add support for other `fp` functions like `trunc`, ...

Let us now if you find some bugs or if you have questions.

**Comment by pablo-angulo** [comment]
BTW, there is another approach to generating parametrized questions, using the new package [pyexams](https://framagit.org/pang/pyexams). It allows to eval code inside any jupyter kernel and creates a fixed set of variants for each question, all of them with the same category so that it is possible to create a exam where each student gets a different variant of each question.

In this way it is possible to use a CAS like Sagemath or sympy to generate the questions and solutions.

Regards

**Comment by nennigb** [comment]
Nice project !
Happy to see that `amc2moodle` parametrized question is useful. Don't hesitate to contact us if you need some other features or if you find some bugs.

Benoit

**Comment by pablo-angulo** [comment]
That project does not use `amc2moodle` parametrized questions, it just generates a few question environments using any `jupyter` kernel. So it is an alternative to `amc2moodle` parametrized questions in the case that you need something more powerful than fpeval.
